src/lc3_vm/virtual_machine.rs

```diff
@@ -0,0 +1,98 @@
+use core::{error, fmt};
+
+const MEMORY_MAX: usize = 1 << 16;
+
+#[derive(Debug)]
+pub enum VMError {
+    LoadProgram(String),
+}
+
+impl fmt::Display for VMError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::LoadProgram(context) => {
+                write!(f, "Failed to load program into memory: {context}")
+            }
+        }
+    }
+}
+
+impl error::Error for VMError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match *self {
+            VMError::LoadProgram(_) => None,
+        }
+    }
+}
+
+pub struct VM {
+    memory: [u16; MEMORY_MAX],
+    r0: u16,
+    r1: u16,
+    r2: u16,
+    r3: u16,
+    r4: u16,
+    r5: u16,
+    r6: u16,
+    r7: u16,
+    pc: u16,
+    cond: u16,
+    count: u16,
+}
+
+#[allow(clippy::upper_case_acronyms)]
+enum ConditionFlags {
+    POS,
+    ZRO,
+    NEG,
+    None,
+}
+
+impl Default for VM {
+    fn default() -> Self {
+        Self {
+            memory: [0; MEMORY_MAX],
+            r0: 0_u16,
+            r1: 0_u16,
+            r2: 0_u16,
+            r3: 0_u16,
+            r4: 0_u16,
+            r5: 0_u16,
+            r6: 0_u16,
+            r7: 0_u16,
+            pc: 0x3000,
+            cond: 0_u16,
+            count: 0_u16,
```

Comment:
Do you need to write `_u16` here? I think the compiler figures out it's a `u16` because of the type specified in `VM` but not sure.

Reply:
Done